uudev.1.md
% uudev(1)
% David Mazières
%

# NAME

uudev - micro-user-udev service

# SYNOPSIS

uudev [-v] [-c conf] \
uudev [-v] -m \
uudev [-v] -p /dev/path

# DESCRIPTION

Micro-udev is a trivial udev-like service that runs commands in the
context of a (non-root) user session when devices are added or removed
from the system.  It fires off shell commands based on properties of
devices assigned by the kernel and system-level udev daemon.

There are three modes.  The main mode (with no options) listens for
device events and fires off rules based on your `uudev.conf`
configuration file.  With `-vm`, uudev runs in monitor mode where it
listens for events and prints what is happening, so you can find
appropriate properties on which to base rules.  With `-vp` uudev
prints the properties of a particular existing device node in the file
system.

Typically you will want to run `uudev -vp` or `uudev -vm` and based on
the output create a `$HOME/.config/uudev.conf` file.  Then test that
it does what you want by running `uudev -v`.  Finally, to start
`uudev`, either just run it (`uudev &`) or start it with systemd:

	systemctl --user restart uudev

Any time you edit the configuration file, you must kill and restart
`uudev`, which you can do by re-running that `systemctl` command.

## OPTIONS

`-v`
:	Increases the level of verbosity.

`-c`
:	Specifies a non-default configuration file location.

`-m`
:	Runs uudev in monitor mode.  This is most useful in conjunction
with `-v` to see all the device properties you can test for.

`-p`
:	Prints the devpath of a device in the file system.  Most useful in
conjunction with `-v`, to get all the device properties you can test
for.

# CONFIGURATION FILE FORMAT

The configuration file consists of a of series stanzas, each of which
starts with a predicate line (beginning `*`) followed by multiple
lines of shell script (not beginning `*`).  Every time there's a
device event matching one or more non-preamble predicates, the shell
scripts of stanzas with matching predicates are concatenated and piped
to `/bin/sh` (with environment variables set for the device event
properties).  Uudev then waits for the shell to complete, so it is
important to fork any long-running programs into the background when
doing this.

The format of a test-line is:

> _TYPE_ {_PROPERTY_ _OP_ `"`_VALUE_`",`}*

_TYPE_ always tarts with `*` and is optionally followed by one or both
of `?` and `!`.  The `!` signifies that the shell code should also be
run whenever `uudev` starts up.  This flag is useful, for instance, if
you want to configure a device such as a bluetooth keyboard when you
first start `uudev` (in case the device is already connected) as well
as whenever the device subsequently connects.  When starting up, `!`
rules are run without any device property environment variables, since
there is no device event.  You can test for this by the lack of an
`ACTION` environment variable.

The `?` flag signifies that the shell script is a "preamble" intended
to be prefixed to other rules.  Preamble shell code will only run if a
non-preamble test-line matches.  This flag is useful to set a shell
variable or define shell functions that apply to all subsequent rules,
or conceivably to filter out certain events by conditionally calling
exit before later stanzas can execute.  For example, if you want to
see all the commands being executed, you can place this at the top of
your file:

    *!?
    set -x

Because the test-line is empty, it matches every device event.  But
because it is marked with a '?', it will not cause a shell to be
spawned for every device event, only ones that also match a
non-preamble line.  However, every time a shell does run it will begin
with `set -x`, which causes the shell to print the commands it is
executing.

The remainder of the test-line consists of tests separated by commas.
Each test checks a _PROPERTY_ against a _VALUE_.  The predicate line
as a whole matches when all tests match (i.e., the tests are logically
anded).  _OP_ can be either `==` or `!=` to check for equality or
inequality.  There is no wildcard pattern matching, though you can
implement more refined tests in shell code launched by a predicate.
Note that _VALUE_ must always be enclosed in double quotes.  If
_VALUE_ is empty, it also matches a missing property, so `!=""` can be
used to test that a property exists.  A backslash (`\`) character in
_VALUE_ escapes the next special character and can be used to insert a
backslash or double-quote character.  There is no way to break a long
predicate line over multiple lines.

The `DEVLINK` property is special.  Since the system udev may create
multiple links to a device, `DEVLINK=="`_VALUE_`"` matches if any of
the device's links is _VALUE_.

# EXAMPLES

To run `xmodmap` every time a keyboard is plugged in, you might use
this configuration:

	*! ACTION=="add", ID_INPUT_KEYBOARD=="1", MAJOR!=""
	xmodmap "$HOME/.mykeymap"

Note `MAJOR!=""` avoids matching devices that get added without any
device node.  This is because sometimes keyboard devices get added
twice, once as a real device and once with `TAGS=:power-switch:`.
Why?  Who knows---the point of uudev is that you don't need to
understand all this systemd stuff and can just create rules by trial
and error.

If you have a Wacom tablet mapped to its own display, the stylus area
will get uncalibrated every time the screen saver comes on.  You can
fix this with the following stanza (if your tablet is xinerama display
`HEAD-2`):

	*! ACTION=="bind", DRIVER=="wacom", HID_UNIQ!=""
	for dev in $(xsetwacom list | sed -ne 's/.*id: \([0-9]*\).*/\1/p'); do
	    xsetwacom set $dev MapToOutput HEAD-2
	done

Why match for non-empty `HID_UNIQ`?  Again, no idea---this is just
trial and error.  There seem to be multiple devices that get added for
the tablet, so to avoid running this command multiple times, or,
worse, running it too early when not all the devices are ready (e.g.,
the touch screen might be ready but not the stylus) testing for an
empty or missing `HID_UNIQ` seems to do the trick.

Say you want to unlock your screen by killing the `xsecurelock`
program every time your phone connects to bluetooth.  (To do this,
obviously you must trust your phone with the `bluetoothctl` `trust`
command, so that it can connect automatically.)  The following stanza
would accomplish this:

	* ACTION=="add", ID_BUS=="bluetooth", NAME=="\"Pixel (AVRCP)\""
	pkill xsecurelock

Again, here you have to use the monitor option (`-vm`) to notice that
udev places the name of your phone in double quotes and adds ` (AVRCP)`.

If you are paranoid, you might want to lock your screen whenever
someone plugs a thumb drive into your computer.  The following rule
would acomplish that:

	* ACTION=="add", ID_BUS=="usb", SUBSYSTEM=="block", ID_DRIVE_THUMB=="1"
	xsecurelock &

Note it is important to put `xsecurelock` into the background with `&`
for two reasons.  First, uudev will block waiting for the command to
complete.  Hence, if the shell script launching `xsecurelock` does not
return immediately, stanzas such as the above bluetooth unlock example
will not run until after `xsecurelock` has exited.  Second, there are
actually multiple matching events when you plug in a thumb drive, so
`xsecurelock` will get run multiple times.  It's smart enough to give
up if the screen is already locked, but if you don't run it in the
background then you will have to type your password several times
before getting your screen back, as each `xsecurelock` will run
immediately after the previous one has exited.

# ENVIRONMENT

uudev just passes environment variables through to the shell code it
executes.  However, it also sets all device properties as environment
variables.

# FILES

The default configuration file location is `$HOME/.config/uudev.conf`.

# BUGS

There are no continuation lines.  There's no way to write a line of
shell code that begins with a `*` character (as might be useful in a
here-document).  There aren't really any comments in the configuration
file (other than above the first predicate line).  Of course, you can
use just shell comments, but they get passed through to the shell.

# SEE ALSO

udev(7)


uudev.cc
#include <poll.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <stdlib.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <libudev.h>

#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <map>
#include <regex>
#include <set>

#include "cleanup.h"

using unique_udev_t = unique_destructor_t<udev_unref>;
using unique_udev_monitor_t = unique_destructor_t<udev_monitor_unref>;
using unique_udev_device_t = unique_destructor_t<udev_device_unref>;

const char confname[] = "uudev.conf";
const char *progname;

static const char *opt_confpath;
int opt_verbose;

static std::string
get_confpath()
{
  std::string ret;
  if (opt_confpath)
    ret = opt_confpath;
  else if (const char *path = getenv("UUDEV_CONF"))
    ret = path;
  else if (const char *conf = getenv("XDG_CONFIG_HOME"))
    ret = std::string(*conf ? conf : ".") + confname;
  else if (const char *home = getenv("HOME"))
    ret = std::string(*home ? home : ".") + "/.config/" + confname;
  else {
    std::cerr << progname << ": no HOME directory and no -c option"
	      << std::endl;
    exit(1);
  }
  if (ret.empty())
    ret = ".";
  return ret;
}

void
waitfd(int fd, short events)
{
  pollfd fds[1];
  memset(fds, 0, sizeof(fds));
  fds[0].fd = fd;
  fds[0].events = events;
  poll(fds, 1, -1);
}

static std::string
upcase(const char *s)
{
  std::string ret;
  while (*s)
    ret += char(toupper(*s++));
  return ret;
}

static void
list_foreach(udev_list_entry *e,
	     std::function<void(std::string, std::string)> cb)
{
  for (int i = 0; e; e = udev_list_entry_get_next(e), ++i) {
    const char *k = udev_list_entry_get_name(e),
      *v = udev_list_entry_get_value(e);
    if (v)
      cb(k, v);
    else
      cb(std::to_string(i), k);
  }
}

static void
dev_foreach(const unique_udev_device_t &dev,
	    std::function<void(std::string, std::string)> cb)
{
  list_foreach(udev_device_get_devlinks_list_entry(dev.get()),
	       [&cb](std::string, std::string v) {
		 cb("DEVLINK", v);
	       });
  list_foreach(udev_device_get_properties_list_entry(dev.get()), cb);
}

struct DevProps {
  std::set<std::string> links_;
  std::map<std::string, std::string> props_;
  DevProps(const unique_udev_device_t &dev);
  bool eq(const std::string &k, const std::string &v) const;
  bool neq(const std::string &k, const std::string &v) const {
    return !eq(k, v);
  }
};
using DevPred = std::function<bool(const DevProps &)>;

DevProps::DevProps(const unique_udev_device_t &dev)
{
  list_foreach(udev_device_get_devlinks_list_entry(dev.get()),
	       [this](std::string, std::string v) {
		 links_.emplace(std::move(v));
	       });
  list_foreach(udev_device_get_properties_list_entry(dev.get()),
	       [this](std::string k, std::string v) {
		 props_.emplace(std::move(k), std::move(v));
	       });
}

bool
DevProps::eq(const std::string &k, const std::string &v) const
{
  if (k == "DEVLINK")
    return links_.find(v) != links_.end() || (links_.empty() && v.empty());
  auto i = props_.find(k);
  return (i != props_.end() && i->second == v) ||
    (i == props_.end() && v.empty());
}

static void
dump(std::ostream &out, const unique_udev_device_t &dev)
{
  if (!dev) {
    out << "NULL" << std::endl;
    return;
  }
  dev_foreach(dev, [&out](std::string k, std::string v) {
		     out << k << "=" << v << std::endl;
		   });
}

static void
dev_setenv(const unique_udev_device_t &dev)
{
  dev_foreach(dev, [](std::string k, std::string v) {
		     setenv(k.c_str(), v.c_str(), 1);
		   });
}

static void
run(const unique_udev_device_t *dev, const std::string cmds)
{
  int fds[2];
  if (pipe(fds) == -1) {
    perror("pipe");
    return;
  }

  pid_t pid = fork();
  if (pid == -1) {
    perror("fork");
    close(fds[0]);
    close(fds[1]);
    return;
  }
  else if (pid == 0) {
    close(fds[1]);
    if (dev)
      dev_setenv(*dev);
    if (fds[0] != 0) {
      dup2(fds[0], 0);
      close(fds[0]);
    }
    execl("/bin/sh", "/bin/sh", nullptr);
    exit(1);
  }
  close(fds[0]);
  if (opt_verbose >= 3) {
    write(0, cmds.data(), cmds.size());
  }
  write(fds[1], cmds.data(), cmds.size());
  close(fds[1]);
  waitpid(pid, nullptr, 0);
}

void
monitor()
{
  unique_udev_t ud (udev_new());
  unique_udev_monitor_t mon (udev_monitor_new_from_netlink(ud.get(), "udev"));
  udev_monitor_enable_receiving(mon.get());
  for (;;) {
    unique_udev_device_t dev(udev_monitor_receive_device(mon.get()));
    if (!dev) {
      waitfd(udev_monitor_get_fd(mon.get()), POLLIN);
      continue;
    }
    std::cout << "* " << udev_device_get_action(dev.get()) << " "
	      << udev_device_get_devpath(dev.get())
	      << " (" << udev_device_get_subsystem(dev.get())
	      << ")" << std::endl;
    if (opt_verbose) {
      dump(std::cout, dev);
      std::cout << "----------------------------------------------"
		<< std::endl;
    }
  }
}

struct Uudev {
  unique_udev_t udev_;
  unique_udev_monitor_t mon_;

  struct Rule {
    DevPred pred_;
    std::string rule_;
    std::string commands_;
    bool immediate_ = false;
    bool preamble_ = false;
    bool parse(const std::string line);
  };
  std::vector<Rule> config_;

  Uudev() : udev_(udev_new()),
	    mon_(udev_monitor_new_from_netlink(udev_.get(), "udev")) {}
  bool parse(std::string path);
  void dumpconf();
  void loop();
};

void
Uudev::loop()
{
  udev_monitor_enable_receiving(mon_.get());

  {
    std::string cmds, debug1;
    bool runit(false);
    for (auto r : config_) {
      if (r.immediate_) {
	cmds += r.commands_;
	if (!r.preamble_) {
	  runit = true;
	  if (opt_verbose >= 1) {
	    debug1 += r.rule_;
	    debug1 += '\n';
	  }
	}
      }
    }
    if (runit) {
      if (opt_verbose >= 1)
	std::cout << "** startup" << std::endl
		  << debug1;
      run(nullptr, cmds);
    }
  }

  for (;;) {
    unique_udev_device_t dev(udev_monitor_receive_device(mon_.get()));
    if (!dev) {
      waitfd(udev_monitor_get_fd(mon_.get()), POLLIN);
      continue;
    }
    DevProps props(dev);
    std::string cmds, debug1;
    bool runit(false);
    for (auto r : config_) {
      if (!r.pred_(props))
	continue;
      if (opt_verbose >= 1) {
	debug1 += r.rule_;
	debug1 += '\n';
      }
      cmds += r.commands_;
      if (!r.preamble_)
	runit = true;
    }
    if (runit) {
      if (opt_verbose >= 1)
	std::cout << "** ACTION==" << udev_device_get_action(dev.get())
		  << ", DEVPATH==\"" << udev_device_get_devpath(dev.get())
		  << "\"" << std::endl
		  << debug1;
      if (opt_verbose >= 2)
	dump(std::cout, dev);
      run(&dev, cmds);
    }
  }
}

void
Uudev::dumpconf()
{
  std::cout << "------------ DUMPING CONFIGURATION -----------" << std::endl;
  for (auto i : config_) {
    std::cout << i.rule_ << std::endl
	      << i.commands_
	      << "----------------------------------------------" << std::endl;
  }
  std::cout << "------------ CONFIGURATION DUMPED ------------" << std::endl;
}

static std::string
dequote(std::string in)
{
  std::string out;
  for (auto c = in.cbegin(); c != in.cend(); ++c) {
    if (*c != '\\') {
      out += *c;
      continue;
    }
    switch (*++c) {
    case 'n':
      out += '\n';
      break;
    case 't':
      out += '\t';
      break;
    default:
      out += *c;
      break;
    }
  }
  return out;
}

static const std::regex rulestart(R"/(^\*[?!]*)/");
static const std::regex rulerx
    (R"/(^\s*(\w+)\s*(==|!=)\s*"(([^"\\]|\\.)*)"\s*)/");
bool
Uudev::Rule::parse(const std::string line)
{
  auto p = line.cbegin(), e = line.cend();
  std::smatch m;
  if (!regex_search(p, e, m, rulestart))
    return false;
  p = m.suffix().first;
  rule_ = line;
  std::string flags(m.str(0));
  immediate_ = flags.find('!') != flags.npos;
  preamble_ = flags.find('?') != flags.npos;
  pred_ = nullptr;

  DevPred ret([](const DevProps &p) { return true; });
  while (regex_search(p, e, m, rulerx)) {
    const std::string k = m.str(1), v = dequote(m.str(3));
    if (m.str(2) == "==")
      ret = [=](const DevProps &p) { return p.eq(k, v) && ret(p); };
    else if (m.str(2) == "!=")
      ret = [=](const DevProps &p) { return p.neq(k, v) && ret(p); };
    else
      return false;

    p = m.suffix().first;
    if (p == e)
      break;
    else if (*p != ',')
      return false;
    p++;
  }
  while (p < e && isspace(*p))
    p++;
  if (p != e)
    return false;
  pred_ = ret;
  return true;
}

bool
Uudev::parse(std::string path)
{
  std::ifstream file(path);
  if (file.fail()) {
    perror(path.c_str());
    return false;
  }

  Rule r;
  bool valid(false);
  for (int lineno = 1;; ++lineno) {
    std::string line;
    std::getline(file, line);
    if (!file || (!line.empty() && line[0] == '*')) {
      if (valid && !r.commands_.empty())
	config_.emplace_back(std::move(r));
      r.commands_.clear();
      if (!file) {
	return file.eof();
      }
    }
    if (line.empty() || line[0] != '*') {
      r.commands_ += line;
      r.commands_ += '\n';
    }
    else if (!(valid = r.parse(line)))
      std::cerr << path << ":" << lineno << ": syntax error" << std::endl;
  }
  return true;
}

void
query_path(const char *path)
{
  struct stat sb;
  if (stat(path, &sb) == -1) {
    perror(path);
    exit(1);
  }
  char type;
  if (S_ISCHR(sb.st_mode))
    type = 'c';
  else if (S_ISBLK(sb.st_mode))
    type = 'b';
  else {
    std::cerr << path << ": not a character or block device" << std::endl;
    exit(1);
  }

  unique_udev_t ud (udev_new());
  unique_udev_device_t dev(udev_device_new_from_devnum(ud.get(), type,
						       sb.st_rdev));
  if (!dev) {
    std::cerr << path << ": could not find device path" << std::endl;
    exit(1);
  }
  if (opt_verbose)
    dump(std::cout, dev);
  else
    std::cout << udev_device_get_devpath(dev.get()) << std::endl;
}

[[noreturn]] static void usage(int exitval = 1);
static void
usage(int exitval)
{
  using namespace std;
  (exitval ? cerr : cout)
    << "usage: " << progname << " [-c conf]" << endl
    << "       " << progname << " [-v] -p /dev/path" << endl
    << "       " << progname << " [-v] -m" << endl;
  exit(exitval);
}

int
main(int argc, char **argv)
{
  if (argc < 1)
    progname = "uudev";
  else if ((progname = strrchr(argv[0], '/')))
    progname++;
  else
    progname = argv[0];

  const char *opt_path = nullptr;
  bool opt_monitor = 0;

  int opt;
  while ((opt = getopt(argc, argv, "hmc:vp:")) != -1) {
    switch (opt) {
    case 'c':
      opt_confpath = optarg;
      break;
    case 'm':
      opt_monitor = true;
      break;
    case 'p':
      opt_path = optarg;
      break;
    case 'v':
      opt_verbose++;
      break;
    case 'h':
      usage(0);
      break;
    default:
      usage(1);
      break;
    }
  }

  if (opt_path)
    query_path(opt_path);
  else if (opt_monitor)
    monitor();
  else {
    Uudev uu;
    if (!uu.parse(get_confpath()))
      exit(1);
    struct sigaction sa;
    memset(&sa, 0, sizeof(sa));
    sa.sa_handler = SIG_IGN;
    sigaction(SIGPIPE, &sa, nullptr);
    if (opt_verbose > 1)
      uu.dumpconf();
    uu.loop();
  }
}
